rs/xnet/payload_builder/src/certified_slice_pool.rs

```diff
@@ -1278,6 +1285,72 @@ impl CertifiedSlicePool {
         }
     }
 
+    pub fn refill_stream_slice_indices(
+        &self,
+    ) -> impl Iterator<Item = (SubnetId, RefillStreamSliceIndices)> {
+        let mut result: BTreeMap<SubnetId, RefillStreamSliceIndices> = BTreeMap::new();
+
+        let pool_slice_stats = {
+            if self.byte_size() > POOL_BYTE_SIZE_SOFT_CAP {
+                // Abort if pool is already full.
+                return result.into_iter();
+            }
+
+            self.peers()
+                // Skip our own subnet, the loopback stream is routed separately.
+                .filter(|&&subnet_id| subnet_id != self.own_subnet_id)
+                .map(|&subnet_id| (subnet_id, self.slice_stats(subnet_id)))
+                .collect::<BTreeMap<_, _>>()
+        };
+
+        for (subnet_id, slice_stats) in pool_slice_stats {
```

Comment:
The whole point of structuring the code this way was to only hold the lock on `PoolRefillTask::pool` while in this top block.

I would either retain that split (and have this function either under `PoolRefillTask` or next to it (all it would require is a reference to an `Arc<Mutex<CertifiedSlicePool>>`); or else refactor it (without this block, `own_subnet_id`, etc.) into a function that acts on a single `SubnetId` at a time.

Reply:
This function is now synchronous and only doing some simple index manipulations => it's supposed to be fast. Do you expect it to make a difference if we release the lock before this for loop and after it?

Reply:
But I'm doing the refactor anyway.

Reply:
Done: https://github.com/dfinity/ic/pull/3376/commits/d2bc30c2e1cecb66024e6bc9654396990a0a9689